arangod/RocksDBEngine/RocksDBBuilderIndex.cpp

```diff
@@ -55,30 +64,296 @@ using namespace arangodb;
-  }
-
- private:
-  DataSourceId _cid;
-};
 
 struct BuilderCookie : public arangodb::TransactionState::Cookie {
   // do not track removed documents twice
   ::arangodb::containers::HashSet<LocalDocumentId::BaseType> tracked;
 };
+
+Result partiallyCommitInsertions(rocksdb::WriteBatchBase& batch,
+                                 rocksdb::DB* rootDB,
+                                 RocksDBTransactionCollection* trxColl,
+                                 std::atomic<uint64_t>& docsProcessed,
+                                 RocksDBIndex& ridx, bool isForeground) {
+  auto docsInBatch = batch.GetWriteBatch()->Count();
+  if (docsInBatch > 0) {
+    rocksdb::WriteOptions wo;
+    rocksdb::Status s = rootDB->Write(wo, batch.GetWriteBatch());
+    if (!s.ok()) {
+      return rocksutils::convertStatus(s, rocksutils::StatusHint::index);
+    }
+  }
+  batch.Clear();
+
+  auto ops = trxColl->stealTrackedIndexOperations();
+  if (!ops.empty()) {
+    TRI_ASSERT(ridx.hasSelectivityEstimate() && ops.size() == 1);
+    auto it = ops.begin();
+    TRI_ASSERT(ridx.id() == it->first);
+
+    auto* estimator = ridx.estimator();
+    if (estimator != nullptr) {
+      if (isForeground) {
+        estimator->insert(it->second.inserts);
+        estimator->remove(it->second.removals);
+      } else {
+        uint64_t seq = rootDB->GetLatestSequenceNumber();
+        // since cuckoo estimator uses a map with seq as key we need to
+        estimator->bufferUpdates(seq, std::move(it->second.inserts),
+                                 std::move(it->second.removals));
+      }
+    }
+  }
+#ifndef USE_SST_INGESTION
+  docsProcessed.fetch_add(docsInBatch, std::memory_order_relaxed);
+#endif
+  return {};
+}
 }  // namespace
 
+IndexCreatorThread::IndexCreatorThread(
+    bool isUniqueIndex, bool isForeground, uint64_t batchSize,
+    std::atomic<uint64_t>& docsProcessed,
+    std::shared_ptr<SharedWorkEnv> sharedWorkEnv, RocksDBCollection* rcoll,
+    rocksdb::DB* rootDB, RocksDBIndex& ridx, rocksdb::Snapshot const* snap,
+    rocksdb::Options const& dbOptions, std::string const& idxPath)
+    : Thread(ridx.collection().vocbase().server(), "IndexCreatorThread"),
+      _isUniqueIndex(isUniqueIndex),
+      _isForeground(isForeground),
+      _batchSize(batchSize),
+      _docsProcessed(docsProcessed),
+      _sharedWorkEnv(std::move(sharedWorkEnv)),
+      _rcoll(rcoll),
+      _rootDB(rootDB),
+      _ridx(ridx),
+      _snap(snap),
+      _trx(transaction::StandaloneContext::Create(_ridx.collection().vocbase()),
+           _ridx.collection(), AccessMode::Type::WRITE),
+      _dbOptions(dbOptions) {
+  if (_isForeground) {
+    _trx.addHint(transaction::Hints::Hint::LOCK_NEVER);
+  }
+  _trx.addHint(transaction::Hints::Hint::INDEX_CREATION);
+  Result res = _trx.begin();
+  if (!res.ok()) {
+    THROW_ARANGO_EXCEPTION(res);
+  }
+  _trxColl = _trx.resolveTrxCollection();
+  if (_isUniqueIndex) {
+    // for later
```

Comment:
Let's put an assertion here, so that this cannot be called with a unique index accidentially.

Reply:
Comment not addressed.

Reply:
done